falcon/request.py

```diff
@@ -571,6 +568,33 @@ def get_header(self, name, required=False):
 
             raise HTTPMissingParam(name)
 
+    def get_header_as_datetime(self, header, required=False):
+        """Return an HTTP header with HTTP-Date values as a datetime.
+
+        Args:
+            name (str): Header name, case-insensitive (e.g., 'Date')
+            required (bool, optional): Set to ``True`` to raise
+                ``HTTPBadRequest`` instead of returning gracefully when the
+                header is not found (default ``False``).
+
+        Returns:
+            datetime: The value of the specified header if it exists,
+                or ``None`` if the header is not found and is not required.
+
+        Raises:
+            HTTPBadRequest: The header was not found in the request, but
+                it was required.
+            HttpInvalidHeader: The header contained a malformed/invalid value.
+        """
+
+        try:
+            http_date = self.get_header(header, required=required)
+            return util.http_date_to_dt(http_date) if http_date is not None \
+                else None
```

Comment:
That was my initial pass, but that caused some unit tests to fail. There are three cases we need to handle:
1. Date header is not set, expect None
2. Date header is set but invalid, expect HTTPInvalidHeader
3. Date header is set and valid, expect parsed datetime

The `get_header()` method does not distinguish between "header is not present" and "header is present but has a value of None or empty string" — both will return None. In the former case, we want to return None. For the latter, we want to raise a corresponding exception.

Here's another stab, but I'm not sure if it's any clearer:
```
        try:
            http_date = self.get_header(header, required=required)
            return util.http_date_to_dt(http_date)
        except TypeError:
            # When the header does not exist and isn't required
            return None
        except ValueError:
            msg = ('It must be formatted according to RFC 1123.')
            raise HTTPInvalidHeader(msg, header)
```